### PR TITLE
Faster gradient square accumulation using hooks

### PR DESCRIPTION
**Pull request by SyphonArch**
This PR fixes two things:

## Speedup

This PR achieves a speedup of more than 20x on an RTX 4090 for calculating the sum of gradients squared.

It does so by using the [`register_hook` method](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html) to have PyTorch calculate the gradient squared on every `loss.backward` invocation.

You simply take the `.grad` attribute of the weights at the very end for the accumulated sum of gradients squared.

I expect this PR to achieve speedups comparable to #6 , but without the extra code/model modification overhead, which is why I'm opening this PR despite #6 already being there.

I did a rough comparison of the resulting values against the original implementation and found that the values are close to identical.
I also tested with [SqueezeLLM](https://github.com/SqueezeAILab/SqueezeLLM) to confirm that the quantized models achieve the same perplexities.

You may do you own tests on this code before integration.

## num_examples bug fix

Fixed an issue where the `num_examples` argument was ineffective for values larger than 128, the default `nsamples` value for the `get_loaders` function.

This moves the sampling logic into the `get_loader` function, and also enables the `tqdm` loading bar to appear, as we don't need to use the `enumerate`.

### CODE REVIEW

**Review by kssteven418 — APPROVED** [review]
Thank you! This looks great and works fine on my end
